src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -0,0 +1,318 @@
+                    document.uri,
+                    document.notebookType
+                );
+                preferredConnection = defaultPythonController?.connection;
+            }
+            if (document.notebookType === JupyterNotebookView && !preferredConnection) {
+                const preferredInterpreter =
+                    !serverId && isPythonNbOrInteractiveWindow && this.extensionChecker.isPythonExtensionInstalled
+                        ? await this.interpreters.getActiveInterpreter(document.uri)
+                        : undefined;
+
+                // Await looking for the preferred kernel
+                ({ preferredConnection } = await this.findPreferredKernelExactMatch(
+                    document.uri,
+                    notebookMetadata,
+                    preferredSearchToken.token,
+                    'useCache',
+                    preferredInterpreter,
+                    serverId
+                ));
+
+                // Also start the search to refresh the cache, don't need to await on this
+                // unless we don't find a match. It will update the cache after running
+                const ignoreCacheFindPreferredPromise = this.findPreferredKernelExactMatch(
+                    document.uri,
+                    notebookMetadata,
+                    preferredSearchToken.token,
+                    'ignoreCache',
+                    preferredInterpreter,
+                    serverId
+                );
+
+                // If we didn't find an exact match in the cache, try awaiting for the non-cache version
+                if (!preferredConnection) {
+                    ({ preferredConnection } = await ignoreCacheFindPreferredPromise);
+                }
+
+                // Send telemetry on looking for preferred don't await for sending it
+                this.sendPreferredKernelTelemetry(
+                    document.uri,
+                    notebookMetadata,
+                    preferredConnection,
+                    preferredInterpreter
+                ).ignoreErrors();
+
+                // If we found a preferred kernel, set the association on the NotebookController
+                if (preferredSearchToken.token.isCancellationRequested && !preferredConnection) {
+                    traceInfo('Find preferred kernel cancelled');
+                    return {};
+                }
+                if (!preferredConnection) {
+                    traceInfoIfCI(
+                        `PreferredConnection not found for NotebookDocument: ${getDisplayPath(document.uri)}`
+                    );
+                    return {};
+                }
+
+                traceInfo(
+                    `PreferredConnection: ${preferredConnection.id} found for NotebookDocument: ${getDisplayPath(
+                        document.uri
+                    )}`
+                );
+
+                const targetController = this.registration.values.find(
```

Comment:
Shouldn't we wait for the controllers to be loaded here just like we do in line 177?
Else we could end in a race condition, where we look for some remote controllers but the promise hasn't competed hence the preferred controller will not be setup.

Reply:
It doesn't do that now, so it's an existing but but it does seem like it should wait right here.

Reply:
Aagreed,

Reply:
No actually it's on purpose. See the comment below. If it isn't loaded yet, we just create it instead.